docs/userguide/src/portingguide/topics/weakref.md

```diff
@@ -0,0 +1,451 @@
+
+Some VMs determine whether an object is finalizable by its type.  In Java, for example, an object is
+finalizable if its `finalize()` method is overridden.  The VM binding can maintain a list of
+finalizable objects, and register instances of such types into that list when they are constructed.
+
+Some VMs can dynamically attach finalizing operations to individual objects after objects are
+created.  The VM binding can maintain a list of objects with attached finalizers, or maintain a
+(weak) hash map that maps finalizable objects to its associated finalizers.
+
+### When to run finalizers?
+
+Depending on the finalizer semantics in different VMs, finalizers can be executed during GC or
+during mutator time after GC.
+
+The VM binding can run finalizers immediately in `Scanning::process_weak_refs` when finding a
+finalizable object unreachable.  Beware that executing finalizers can be time-consuming.  The VM
+binding can creating work packets and let each work packet process a part of all finalizable
+objects.  In this way, multiple GC workers can process finalizable objects in parallel.  The
+`Scanning::process_weak_refs` function is executed in the `VMRefClosure` stage, so the created work
+packets shall be added to the same bucket.
+
+If the finalizers should be executed after GC, the VM binding should enqueue such jobs to
+VM-specific queues so that they can be picked up by mutator threads after GC.
+
+### Reading the body of dead object
+
+In some VMs, finalizers can read the fields in dead objects.  Such fields usually include
+information needed for cleaning up resources held by the object, such as file descriptors and
+pointers to memory not managed by GC.
+
+`Scanning::process_weak_refs` is executed in the `VMRefClosure` stage, which happens after computing
+transitive closure, but before any object has been released (which happens in the `Release` stage).
+This means the body of all objects, live or dead, can still be accessed during this stage.
+
+Therefore, there is no problem reading the object body if the VM binding executes finalizers
+immediately in `process_weak_refs`, or in created work packets in the `VMRefClosure` stage.
+
+However, if the VM needs to execute finalizers after GC, it will be a problem because the object
+will have been reclaimed, and memory of the object will have been overwritten by other objects.  In
+this case, the VM will need to "resurrect" the dead object.
+
+### Resurrecting dead objects
+
+Some VMs, particularly the Java VM, executes finalizers during mutator time.  The dead finalizable
+objects must be brought back to life so that they can still be accessed after the GC.
+
+The `Scanning::process_weak_refs` has an parameter `tracer_context: impl ObjectTracerContext<VM>`.
+This parameter provides the necessary mechanism to resurrect objects and make them (and their
+descendants) live through the current GC.  The typical use pattern is:
+
+```rust
+impl<VM: VMBinding> Scanning<VM> for VMScanning {
```

Comment:
It is generally preferrable to put the code into a test, and include the snippets in the test in the doc. In such cases, the code will be always checked by the tests and the CI. Maybe we cannot run `process_weak_refs` in our mock tests, but we can still make sure that it can be compiled. 

See here for examples: https://github.com/mmtk/mmtk-core/blob/051bc7470feef915c445305301e6113f86d3957b/docs/userguide/src/portingguide/perf_tuning/alloc.md?plain=1#L36

Reply:
I now use a mock test to contain the example code.  This can ensure it compiles.  But we still can't actually run it in mock tests because we can't construct a `GCWorker` object.